protocol/src/test/scala/io/mediachain/protocol/JournalBlockGenerators.scala

```diff
@@ -0,0 +1,193 @@
+package io.mediachain.protocol
+
+object JournalBlockGenerators {
+  import io.mediachain.protocol.Datastore._
+  import org.scalacheck._
+  import DataObjectGenerators._
+
+
+  case class JournalBlockWithDatastore(block: JournalBlock, datastore: InMemoryDatastore)
+  case class BlockChainWithDatastore(blockChain: List[JournalBlock], datastore: InMemoryDatastore)
+
+  /**
+    * Make a generator that returns a chain cell for the given `CanonicalRecord`
+    *
+    * @param canonical either an `Entity` or `Artefact` to generate a cell for
+    * @return depending on the type of `CanonicalRecord` given, a generator for
+    *         either a subtype of `EntityChainCell` or `EntityArtefactCell`.
+    *         Note that the `chain` field of the generated cell will be
+    *         set to `None`
+    *
+    */
+  def genCellFor(canonical: CanonicalRecord)
+  : Gen[ChainCell] =
+    canonical match {
+      case e: Entity => genEntityUpdateCell(e, genNilReference)
+      case a: Artefact => genArtefactUpdateCell(a, genNilReference)
+    }
+
+
+  /**
+    * Takes a list of `ChainCell`s with nil `chain` references,
+    * and joins them together so that cells which refer to the same
+    * `CanonicalRecord` are "consed" together via their `chain` references.
+    *
+    * @param cells a list of `ChainCell`s with nil `chain` references
+    * @return the input list, with the `chain` field of each cell updated
+    *         so that it points to the previous cell that refers to the same
+    *         `CanonicalRecord`.
+    */
+  def consChainCells(cells: List[ChainCell]): List[ChainCell] = {
+    import collection.mutable.{Map => MMap, MutableList => MList}
+    val chainHeads: MMap[Reference, ChainCell] = MMap()
+    val consed: MList[ChainCell] = MList()
+
+    cells.foreach { c =>
```

Comment:
not super important but this could be a stateless fold w/ accumulator `(Map..., List)` where u just `._2` after folding to pop out your result

or a recursive fn in the `State` monad from cats (prob easiest)